### PR TITLE
fix: add sfs turbo csi controller api

### PR DESCRIPTION
**Pull request by deer-hang**
1、adjust tool library
2、add controller server api

### CODE REVIEW

**Comment by chengxiangdong** [comment]
pls remove the empty file `pkg/sfsturbo/services/computes.go`

File: pkg/sfsturbo/services/volumes.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+				"Failed to query share %s when wait share available: %v", shareID, err)
+		}
+		log.V(4).Infof("[DEBUG] query share detail: %v", share)
+		if share.Status == shareAvailable {
+			return true, nil
+		}
+		if share.Status == shareCreating {
+			return false, nil
+		}
+		return false, status.Error(codes.Internal, "created share status is not available")
```
Comment (chengxiangdong) [inline pkg/sfsturbo/services/volumes.go:55]: how about the following:
`status.Errorf(codes.Internal, "created share status is not available : %s", share.Status)`

File: pkg/sfsturbo/services/volumes.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	})
+}
+
+func CreateShare(client *golangsdk.ServiceClient, createOpts *shares.CreateOpts) (*shares.TurboResponse, error) {
+	createOpts.Description = shareDescription
+	share, err := shares.Create(client, createOpts).Extract()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to create share, err: %v", err)
```
Comment (chengxiangdong) [inline pkg/sfsturbo/services/volumes.go:63]: 当资源不存在时SDK返回err（是ErrDefault404对象），有时候资源不存在并不需要报错返回，我们会用common.IsNotFound进行判断。
参考这个写法[Link](https://github.com/huaweicloud/huaweicloud-csi-driver/blob/master/pkg/evs/services/volumes.go#L45)

File: pkg/sfsturbo/services/volumes.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	}
+	return share, nil
+}
+
+func DeleteShareCompleted(client *golangsdk.ServiceClient, shareID string) error {
+	if err := DeleteShare(client, shareID); err != nil {
+		if common.IsNotFound(err) {
+			log.V(4).Infof("share %s not found, assuming it to be already deleted", shareID)
```
Comment (chengxiangdong) [inline pkg/sfsturbo/services/volumes.go:71]: the description "assume" is misleading, pls revise

File: pkg/sfsturbo/services/volumes.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+		} else {
+			return status.Errorf(codes.Internal, "Failed to delete share, err: %v", err)
+		}
+	}
+	return WaitForShareDeleted(client, shareID)
+}
+
+func DeleteShare(client *golangsdk.ServiceClient, shareID string) error {
+	return shares.Delete(client, shareID).ExtractErr()
+
```
Comment (chengxiangdong) [inline pkg/sfsturbo/services/volumes.go:82]: remove the empty line

File: pkg/sfsturbo/services/volumes.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+// WaitForShareDeleted delete share from sfs
+func WaitForShareDeleted(client *golangsdk.ServiceClient, shareID string) error {
+	return common.WaitForCompleted(func() (bool, error) {
+		share, err := GetShare(client, shareID)
+		if err != nil {
+			return false, status.Errorf(codes.Internal,
+				"Failed to query share %s when wait share deleted: %v", shareID, err)
+		}
+		log.V(4).Infof("[DEBUG] query share detail: %v", share)
+		if share.Status == shareDeleted {
```
Comment (chengxiangdong) [inline pkg/sfsturbo/services/volumes.go:94]: In fact, the status of deleted resources cannot be found

File: pkg/sfsturbo/services/volumes.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	return shares.Delete(client, shareID).ExtractErr()
+
+}
+
+// WaitForShareDeleted delete share from sfs
+func WaitForShareDeleted(client *golangsdk.ServiceClient, shareID string) error {
+	return common.WaitForCompleted(func() (bool, error) {
+		share, err := GetShare(client, shareID)
+		if err != nil {
+			return false, status.Errorf(codes.Internal,
```
Comment (chengxiangdong) [inline pkg/sfsturbo/services/volumes.go:90]: Should judge whether the resource exists according to `err'

**Comment by chengxiangdong** [comment]
Each function in services has parameter `client *golangsdk.ServiceClient`, which is unreasonable, how to call sdk should be wrapped in services. 
There is no need to get `client, err := cs.Driver.cloud.SFSTurboV1Client()` in the controller or nodeServer, just pass the `cs.Driver.cloud` to services functions.

File: pkg/sfsturbo/services/volumes.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	}
+	log.V(4).Infof("[DEBUG] create share response detail: %v", turboResponse)
+	err = WaitForShareAvailable(client, turboResponse.ID)
+	if err != nil {
+		return nil, err
+	}
+	return turboResponse, nil
+}
+
+// WaitForShareAvailable create new share from sfsturbo
```
[the remark this answers is not in the document]
Comment (deer-hang) [reply pkg/sfsturbo/services/volumes.go:40]: Description has been removed

File: pkg/sfsturbo/services/volumes.go (hunk 1 above)
Comment (deer-hang) [reply pkg/sfsturbo/services/volumes.go:55]: done

File: pkg/sfsturbo/services/volumes.go (hunk 3 above)
Comment (deer-hang) [reply pkg/sfsturbo/services/volumes.go:71]: done

File: pkg/sfsturbo/services/volumes.go (hunk 4 above)
Comment (deer-hang) [reply pkg/sfsturbo/services/volumes.go:82]: done

File: pkg/sfsturbo/services/volumes.go (hunk 5 above)
Comment (deer-hang) [reply pkg/sfsturbo/services/volumes.go:94]: done

File: pkg/sfsturbo/services/volumes.go (hunk 6 above)
Comment (deer-hang) [reply pkg/sfsturbo/services/volumes.go:90]: done
use GetVolume() to judge whether the resource exists

File: examples/sfsturbo-csi-plugin/kubernetes/sc.yaml [hunk 8]
```diff
@@ -4,3 +4,6 @@ metadata:
   name: sfsturbo.csi.huaweicloud.com
 provisioner: sfsturbo.csi.huaweicloud.com
 reclaimPolicy: Delete
+parameters:
+  # shareType is optional, should be 'STANDARD' or 'PERFORMANCE', default value is 'STANDARD'
```
[the remark this answers is not in the document]
Comment (deer-hang) [reply examples/sfsturbo-csi-plugin/kubernetes/sc.yaml:8]: done

File: examples/sfsturbo-csi-plugin/kubernetes/sc.yaml [hunk 9]
```diff
@@ -4,3 +4,6 @@ metadata:
   name: sfsturbo.csi.huaweicloud.com
 provisioner: sfsturbo.csi.huaweicloud.com
 reclaimPolicy: Delete
+parameters:
+  # shareType is required, should be 'STANDARD' or 'PERFORMANCE', default to 'STANDARD'
```
Comment (chengxiangdong) [inline examples/sfsturbo-csi-plugin/kubernetes/sc.yaml:8]: `defaults to` not `default to`

Comment (deer-hang) [reply examples/sfsturbo-csi-plugin/kubernetes/sc.yaml]: done

File: pkg/sfsturbo/services/volumes.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+	}
+	log.V(4).Infof("[DEBUG] create share response detail: %v", protosanitizer.StripSecrets(turboResponse))
+	err = WaitForShareAvailable(c, turboResponse.ID)
+	if err != nil {
+		return nil, err
+	}
+	return turboResponse, nil
+}
+
+const (
```
Comment (chengxiangdong) [inline pkg/sfsturbo/services/volumes.go:43]: pls put constant definitions at the top

File: pkg/sfsturbo/services/volumes.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+	}
+	share, err := shares.Create(client, createOpts).Extract()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to create share, err: %v", err)
+	}
+	return share, nil
+}
+
+func DeleteShareCompleted(c *config.CloudCredentials, shareID string) error {
+	if _, err := GetShare(c, shareID); err != nil {
```
Comment (chengxiangdong) [inline pkg/sfsturbo/services/volumes.go:88]: there should be no need to query, the API will return 404 when the resource does not exist

File: pkg/sfsturbo/services/volumes.go (hunk 10 above)
Comment (deer-hang) [reply pkg/sfsturbo/services/volumes.go:43]: done

File: pkg/sfsturbo/services/volumes.go (hunk 11 above)
Comment (deer-hang) [reply pkg/sfsturbo/services/volumes.go:88]: For DeleteShare, 400 will be returned when the resource does not exist.
Under the same conditions, GetShare will return 404

File: pkg/sfsturbo/controllerserver.go [hunk 12]
```diff
@@ -18,179 +18,425 @@ package sfsturbo
...
+	name := req.GetName()
+	capacityRange := req.GetCapacityRange()
+	if err := createVolumeValidation(name, capacityRange); err != nil {
+		return nil, err
+	}
+	sizeInGiB := int(utils.RoundUpSize(capacityRange.GetRequiredBytes(), common.GbByteSize))
+	// 500 ~ 32768
+	if sizeInGiB < minSizeInGiB {
+		sizeInGiB = minSizeInGiB
+	} else if sizeInGiB > maxSizeInGiB {
```
Comment (chengxiangdong) [inline pkg/sfsturbo/controllerserver.go:69]: exceeding the maximum capacity should return an error, do not create a small one

File: pkg/sfsturbo/controllerserver.go [hunk 13]
```diff
@@ -18,179 +18,425 @@ package sfsturbo
...
+				"SFS-Turbo name: %s already exists with different attributes", createOpts.Name)
+		}
+	}
+	return nil, nil
+}
 
-	volID := req.GetVolumeId()
-	if len(volID) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "DeleteVolume Volume ID must be provided")
+func getAZFromTopology(requirement *csi.TopologyRequirement) string {
```
Comment (chengxiangdong) [inline pkg/sfsturbo/controllerserver.go:173]: pls use `common.GetAZFromTopology` instead

File: pkg/sfsturbo/controllerserver.go (hunk 12 above)
Comment (deer-hang) [reply pkg/sfsturbo/controllerserver.go:69]: done

File: pkg/sfsturbo/controllerserver.go (hunk 13 above)
Comment (deer-hang) [reply pkg/sfsturbo/controllerserver.go:173]: done

File: pkg/sfsturbo/controllerserver.go [hunk 14]
```diff
@@ -18,179 +18,406 @@ package sfsturbo
...
+		}
+	}
+
+	if err = services.ExpandShareCompleted(cloud, volumeID, sizeInGiB); err != nil {
+		return nil, err
+	}
+	log.Infof("Successfully resized volume %v to size %v", volumeID, sizeInGiB)
+	return &csi.ControllerExpandVolumeResponse{
+		CapacityBytes:         int64(sizeInGiB),
+		NodeExpansionRequired: true,
```
Comment (chengxiangdong) [inline pkg/sfsturbo/controllerserver.go:410]: modify to false

**Pull request merged**